src/nvim/os/stdpaths.c

```diff
@@ -0,0 +1,100 @@
+#include "nvim/os/os.h"
+#include "nvim/strings.h"
+#include "nvim/path.h"
+#include "nvim/garray.h"
+
+typedef enum {
+  kConfigHome,
+  kDataHome,
+  kCacheHome,
+  kRuntimeDir,
+  kConfigDirs,
+  kDataDirs,
+} XDGDirType;
+
+static const char *xdg_env_vars[] = {
+  [kConfigHome] = "XDG_CONFIG_HOME",
+  [kDataHome] = "XDG_DATA_HOME",
+  [kCacheHome] = "XDG_CACHE_HOME",
+  [kRuntimeDir] = "XDG_RUNTIME_DIR",
+  [kConfigDirs] = "XDG_CONFIG_DIRS",
+  [kDataDirs] = "XDG_DATA_DIRS",
+};
+
+static const char *const xdg_defaults[] = {
+  // Windows, Apple stuff are just shims right now
+#ifdef WIN32
+  // Windows
+#elif APPLE
+  // Apple (this includes iOS, which we might need to handle differently)
+  [kConfigHome] = "~/Library/Preferences",
+  [kDataHome] = "~/Library/Application Support",
+  [kCacheHome] = "~/Library/Caches",
+  [kRuntimeDir] = "~/Library/Application Support",
+  [kConfigDirs] = "/Library/Application Support",
+  [kDataDirs] = "/Library/Application Support",
+#else
+  // Linux, BSD, CYGWIN
+  [kConfigHome] = "~/.config",
+  [kDataHome] = "~/.local/share",
+  [kCacheHome] = "~/.cache",
+  [kRuntimeDir] = "",
+  [kConfigDirs] = "/etc/xdg/",
+  [kDataDirs] = "/usr/local/share/:/usr/share/",
+};
+#endif
+
+static const char *get_xdg(XDGDirType idx)
+{
+  const char *env = xdg_env_vars[idx];
+  const char *fallback = xdg_defaults[idx];
+
+  const char *ret = os_getenv(env);
+  if (!ret && fallback) {
+    ret = (const char *)expand_env_save((char_u *)fallback);
+  }
+
+  return ret;
```

Comment:
This returns statically allocated memory when `os_getenv(env)` returns non-null, and heap-allocated memory when it returns null and `fallback` isn't null. How does the caller know whether the return value requires an `xfree()`? We could add a parameter, `bool *must_free`, but it may be simpler to always return allocated memory.

```c
const char *ret = os_getenv(env);
if (ret) {
  ret = xstrdup(ret);
} else if (fallback) {
  // ... same as before
}
```